Add the ability to customise the CocoaPods-Try via a .cocoapods.yml

orta opened the pull request:
Adds the ability to run custom pre installation steps for `pod try` via a `.cocoapods.yml`. Should allow for fixing #38 and #33

kylef commented on spec/command/try_settings_spec.rb:
```diff
@@ -0,0 +1,71 @@
+    it 'returns an instance with the right defaults when there are no yml settings files' do
+      Dir.mktmpdir do |dir|
+        yaml = <<eos
+          try_pre_install:
```
Why not use something like this:

```yaml
try:
  install:
    pre:
      - pod install
      - git submodule init
  workspace: ORStackView.xcworkspace
```

That way all plugin related things are grouped at the root.

orta replied:
I like this idea :+1:

orta commented:
Alright, I've also updated the `Gemfile.lock` and removed one of the new checks in RuboCop - it was only applying to the tests.

Ran it with against a fork of a pod I had laying around. `bundle exec pod try https://github.com/orta/AMTagListView --verbose` Works :+1:

kylef commented on CHANGELOG.md:
```diff
@@ -1,5 +1,26 @@
+##### Enhancements
+
+* A CocoaPods can now declare their own pre-install commands, and prefer a project.
+  To use this, add a `.cocoapods.yml` file to your repo. The yaml file should have
```
file in the **root** of your repository.?

orta replied:
:+1:

kylef commented on lib/pod/try_settings.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+
+    # If we need to run commands from pod-try we should let the users know
+    # what is going to be running on their device.
```
:+1: awesome!

kylef commented on lib/pod/try_settings.rb:
```diff
@@ -0,0 +1,47 @@
+
+      settings = YAML.load(File.read(settings_path))
+      try_settings = TrySettings.new
+      try_settings.pre_install_commands = Array(settings['try']['install']['pre'])
```
Wouldn't this crash if we have a settings file and it doesn't have an install pre setting?

orta replied:
```
  - handles not including system commands in the pre-install [FAILED]
  - handles not including the try in the yaml [FAILED]
```

Yep, great catch @kylef - thanks. Will fix,

alexito4 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,26 @@
+
+##### Enhancements
+
+* A CocoaPods can now declare their own pre-install commands, and prefer a project.
```
Shouldn't this say "A Pod..."? :question:

orta replied:
Any, and any repo too, changed it.

segiddins commented on .rubocop.yml:
```diff
@@ -1,3 +1,6 @@
  - .rubocop_cocoapods.yml
+
+Metrics/ModuleLength:
+  Enabled: false
```
iirc this is already disabled in the standard CocoaPods config?

orta replied:
Not in try, I can pull in  anew one from CP ( I updated the rubocop gem )

segiddins commented on CHANGELOG.md:
````diff
@@ -1,5 +1,26 @@
+  ```
+
+  [Orta Therox](https://github.com/orta)
+  Fixes [cocoapods-try#33](https://github.com/CocoaPods/cocoapods-try/issues/33)
````
we normally only link to the issue, and don't prefix the number when its from the same repo

segiddins commented on lib/pod/try_settings.rb:
```diff
@@ -0,0 +1,53 @@
+        commands = pre_install_commands.length > 1 ? 'commands' : 'command'
+        UI.puts "In order to try this pod, CocoaPods-Try needs to run the following #{commands}:"
+        pre_install_commands.each { |command| UI.puts " - #{command}" }
+        UI.puts "\nPress return to run these #{commands}, or press `cmd + c` to stop trying this pod."
```
control-c sends SIGINT

orta replied:
can we capture this and elegantly end?

segiddins replied:
hm? I mean it's control-c, not command-c.

orta replied:
:D

segiddins commented on lib/pod/try_settings.rb:
```diff
@@ -0,0 +1,53 @@
+    def run_pre_install_commands(prompt)
+      if pre_install_commands
+        prompt_for_permission if prompt
+        pre_install_commands.each { |command| system command }
```
Let's do what we do for the prepare command here?

```ruby
Executable.execute_command('bash', ['-e', command], true)
```

segiddins commented:
Excellent work @orta ! Awesome test coverage :)

segiddins commented on lib/pod/try_settings.rb:
```diff
@@ -0,0 +1,53 @@
+      settings_path = Pathname.new(path) + '.cocoapods.yml'
+      return TrySettings.new unless File.exist? settings_path
+
+      settings = YAML.load(File.read(settings_path))
```
use `YAMLHelper.load_file(settings_path)`

segiddins commented on lib/pod/try_settings.rb:
```diff
@@ -0,0 +1,53 @@
+      try_settings = TrySettings.new
+      return try_settings unless settings['try']
+
+      if settings['try']['install']
```
no need for the `if` -- `Array(nil)` returns `[]`

orta replied:
but aren't these dicts?

segiddins commented on spec/command/try_settings_spec.rb:
```diff
@@ -0,0 +1,105 @@
+
+    it 'returns an instance with the right defaults when there are no yml settings files' do
+      Dir.mktmpdir do |dir|
+        yaml = <<eos
```
`YAML` instead of `eos` so we can get syntax highlighting?

segiddins commented:
:+1: feel free to merge once the comments are addressed.

segiddins commented:
:shipit:

orta commented:
:tada:
